_content/devices/bytes.md

```diff
@@ -158,38 +166,73 @@ int myVal = ((int)(payload[0]) << 8)
 
 ### 4. Shift bits
 
-If the range of expected values is bigger than 65536 we can use the same trick. The only difference is that we have to manually shift bits when we encode on Arduino, just like we did in the payload function.
+If the range of expected values is bigger than 65536 we can use the same trick. The only difference is that we have to manually shift bits when we encode on Arduino, just like we did in the payload function, and we need to be careful while decoding in JavaScript.
 
-Let's say we need to encode a [long](https://www.arduino.cc/en/Reference/Long) which uses 4 bytes for a range up to 4294967296.
+Let's say we need to encode an unsigned [long](https://www.arduino.cc/en/Reference/Long) which uses 4 bytes for a range of 0 up to 4294967296.
 
 Encode (Arduino):
 
 ```c
-long lng = 200000L;
+unsigned long lng = 200000L;
 byte payload[4];
 payload[0] = (byte) ((lng & 0xFF000000) >> 24 );
 payload[1] = (byte) ((lng & 0x00FF0000) >> 16 );
 payload[2] = (byte) ((lng & 0x0000FF00) >> 8  );
 payload[3] = (byte) ((lng & 0X000000FF)       );
 ```
 
+For Arduino you will often see that the masks such as `0x00FF0000`, and the type casts to `(byte)`, are not used. That works when `payload` is defined as an array of `byte` as that will really only fit 8 bits, where only the 8 least significant bits are stored and all others are simply discarded. But it never hurts to be explicit, and other platforms might surely need the masks and casts.
+
 Decode (payload functions):
 
 ```js
-decoded.myVal = ((long)(bytes[0]) << 24)
-              + ((long)(bytes[1]) << 16)
-              + ((long)(bytes[2]) << 8)
-              + ((long)(bytes[3]));
+// For unsigned 32 bits number
+decoded.myVal = bytes[0] * 0x1000000
+              + bytes[1] * 0x10000
+              + bytes[2] * 0x100
+              + bytes[3];
```

Comment:
Let's leave this `>>` and `<<` consistent.

Reply:
That would be great, but: how?

JavaScript's bitwise operators will yield wrong results for **unsigned** 32 bits integers if applying bitwise operators to the MSB. Like, e.g., 0xF0000000 is unsigned 4026531840, but would decode to signed -268435456 when using bitwise operators on the MSB. (Currently the documentation only mentions signed numbers _after_ this very use case; maybe we should change the order.)

```js
function Decoder(bytes, port) {
  return {
    int32ok_a: bytes[0] * 0x1000000 + bytes[1] * 0x10000 + bytes[2] * 0x100 + bytes[3],
    int32ok_b: bytes[0] * (1<<24) + bytes[1] * (1<<16) + bytes[2] * (1<<8) + bytes[3],
    // It's just the MSB that needs special care; the following works too, but: ugly
    int32ok_c: bytes[0] * (1<<24) + (bytes[1]<<16 | bytes[2]<<8 | bytes[3]),

    // Things go south when using bitwise operators for the MSB:
    int32wrong_a: bytes[0]<<24 | bytes[1]<<16 | bytes[2]<<8 | bytes[3],
    int32wrong_b: (bytes[0]<<24) + (bytes[1]<<16) + (bytes[2]<<8) + bytes[3],
    int32wrong_c: bytes[0] * (1<<24) | (bytes[1]<<16 | bytes[2]<<8 | bytes[3]),
  }
}
```

...yields the following for 0xF00000:

```json
{
  "int32ok_a": 4026531840,
  "int32ok_b": 4026531840,
  "int32ok_c": 4026531840,
  "int32wrong_a": -268435456,
  "int32wrong_b": -268435456,
  "int32wrong_c": -268435456
}
```

Reply:
I banished this JS stuff so I didn't realize it's tricky to do this with bitshifts, but it is possible;

```js
function Decoder(bytes, port) {
  return {
    val: ((bytes[0] & 0x80 ? -256 : 0) + bytes[0]) << 24 | bytes[1] << 16 | bytes[2] << 8 | bytes[3],
  }
}
```

Reply:
BTW, `0xF0000000` as `int32` is -16777216.

Reply:
Your version is also giving me a negative number, for `F0000000`:

> ![unsigned int showing as signed](https://user-images.githubusercontent.com/1497002/53090625-4c751080-3510-11e9-901c-81ad6f2cc2b2.png)

Also,

> BTW, `0xF0000000` as `int32` is -16777216.

Uh, not in two's complement? How did you calculate that?

Reply:
Yes, an `int32`, so a signed 32-bit integer; the first being the sign. That's what it's supposed to be right? Or what do you mean?

Sorry for that last comment, I meant `0xFF000000`.

Reply:
Up to you if you want to use this bitshift approach, or the multiplication. I personally find the bitshift approach with explicit sign more clear.

Reply:
The change in the Decoder I suggested is for the unsigned example, which in the text for the Arduino code mentioned "up to 4294967296", hence is unsigned. (Actually, "up to" there means "not including"; alternatively "0 through 4294967295" might be preferred. Oh well.)

Maybe we should first give a signed example (plain shifting, no thinking required, JavaScript matches Arduino/C/C++ for that). Still then, at some point the unsigned Decoder should be mentioned too, I guess... Or maybe we should simply give _both_ decoders for the 32 bits example (hence: also mention the signed example, even though the section "How to send negative numbers?" follows _after_ "How to send big numbers?").

Anyway, an additional [zero-fill right shift](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators#%3E%3E%3E_(Zero-fill_right_shift)) (a.k.a. unsigned right shift) `>>> 0` makes JavaScript change its internal type too, but needs parentheses:

```js
// For signed 32 bits integer, -2147483648 through 2147483647
decoded.myVal = bytes[0] << 24
              | bytes[1] << 16
              | bytes[2] << 8
              | bytes[3];

// For unsigned 32 bits integer, 0 through 4294967295
decoded.myVal = (bytes[0] << 24
              | bytes[1] << 16
              | bytes[2] << 8
              | bytes[3]) >>> 0;
```

It surely looks like the other examples though! :-)

Reply:
Ah, now I get it. JavaScript _does_ account for the sign. It's always hacky ...

Reply:
I agree with Johan that bitshifts are more clear than multiplication. In the embedded world bitshifts are also supposed to be faster than multiplication and division - maybe the same is true for Javascript?

Having both examples for signed and unsigned looks like the best approach to me. Seeing as this paragraph talks about unsigned, only have the unsigned example here. A little lower down you have the signed example.